play-json/shared/src/main/scala/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -717,8 +717,8 @@ import scala.reflect.macros.blackbox
           //   invoke the nullableWithDefault version
           (isOption, defaultValue) match {
             case (true, Some(v)) =>
-              val c = TermName(s"${methodName}NullableWithDefault")
-              q"$jspathTree.$c($v)($impl)"
+              val c = TermName(s"${methodName}HandlerWithDefault")
+              q"$config.optionHandlers.$c($jspathTree, $v)($impl)"
```

Comment:
The read/write behavior we want is defined in the `OptionHandler`. Previously the macro would expand to `formatNullableWithDefault` which is effectively hard-coding `OptionHandlers.Default`.

Reply:
I still don't get it because both implementations of `readHandlerWithDefault` are `jsPath.readNullableWithDefault(defaultValue)`.
So I guess I don't understand the fix.
I probably need pull this PR down and look at it to understand it.

Reply:
Yes, when the macro is used for methodName=read, it will evaluate to `readNullableWithDefault` for both `OptionHandler` implementations.

The problem is when the macro is used for methodName=format. `formatNullableWithDefault` is the combination of `readNullable` + `writeNullable`. There's no `format` equivalent for `readNullable` + `writeOptionWithNull`. So we'd have to inline some boilerplate into the macro for the `format` case. I put the boilerplate into `OptionHandler` instead.

Reply:
Ah, I'm starting to see how the pieces fall into place.

I'd like to, if possible, avoid the new methods on `OptionHandlers`, so I'll test that out quickly tomorrow (it's getting late here).

Reply:
OK I now understand.